llvm/lib/Analysis/InstructionSimplify.cpp

```diff
@@ -4316,12 +4309,30 @@ static Value *simplifyWithOpReplaced(Value *V, Value *Op, Value *RepOp,
   if (isa<FreezeInst>(I))
     return nullptr;
 
+  SmallVector<std::pair<Value *, Value *>> ValidReplacements{};
+  for (const auto &OpAndRepOp : Ops) {
+    // We cannot replace a constant, and shouldn't even try.
+    if (isa<Constant>(OpAndRepOp.first))
+      return nullptr;
+
+    // For vector types, the simplification must hold per-lane, so forbid
+    // potentially cross-lane operations like shufflevector.
+    if (OpAndRepOp.first->getType()->isVectorTy() &&
+        !isNotCrossLaneOperation(I))
+      continue;
```

Comment:
I think we should just `return nullptr` here as well. Handling this more precisely is not worth constructing the separate vector. (In practice all the types are the same anyway.)

Reply:
Think we need to create second vec anyways for `const` case (which should be continue as opposed to `return`).

Reply:
The const case should also return (as you already do).

Reply:
I think thats a mistake. If say we have `(select (icmp eq (or x, y), 0), (add x, 1), ...)`, we don't want to bail on simplifying the `add` just because its rhs is constant.

Reply:
That's not what this check does. It protects against doing something like "replace 0 with x".

Reply:
But if we bail when any op is constant we will end up bailing on the entire simplification no?

Reply:
Oh I see, yeah I had it the other way around. I guess we are guarding against `(icmp eq (or x, 1), 0)` which will get simplification elsewhere. Will change.